### ISSUE TITLE
Make it harder to leak generated credentials

### ISSUE THREAD

**Issue by matthieuprat** [google-github-actions/auth#275]
### TL;DR

Early this week, we noticed that we were leaking generated credentials (`gha-creds-*.json` files) from a GitHub Actions job.

We've been thinking about a couple of ways to prevent this from happening again for us and others:
- Store the credentials in environment variables (which is apparently [what AWS is doing](https://github.com/aws-actions/configure-aws-credentials))
- Store the credentials file in another directory (still accessible to Docker-based actions)

I understand that the first option is a bit more effort as it would require upgrading actions which consume those credentials so I'm detailing below the second option.

### Detailed design

Ideally, we would store the credentials file in the `RUNNER_TEMP` directory but this directory is not available to Docker-based actions. However, the `$RUNNER_TEMP/_github_home` directory [is](https://github.com/actions/runner/blob/0e7ca9aedb2cf4ed792303c86b0598a43e773aa1/src/Runner.Worker/Handlers/ContainerActionHandler.cs#L183) and seems like an appropriate place to stash credentials.

Thus, my suggestion would be to store the credentials file in that directory instead.

### Additional information

The underlying issue is well documented in the [README](https://github.com/google-github-actions/auth#prerequisites) and in the [TROUBLESHOOTING](https://github.com/google-github-actions/auth/blob/main/docs/TROUBLESHOOTING.md#dirty-git-or-bundled-credentials) section. This helped us put together a fix but I'm still concerned that we (and others) will hit this issue again.

I'm happy to follow up with a pull request if the strategy above sounds sensible.

**Comment by sethvargo** [comment]
Hi @matthieuprat 

I think there's a few things to call out here:

1. We've explored environment variables and alternative file paths, but they all come with trade-offs, particularly around self-hosted runner threat models. Last time we tried to "fix" this, we accidentally broke all Docker-based actions.

    I understand a subdirectory of RUNNER_TEMP is currently available to docker-based actions, but that's not documented anywhere. We've been burnt by reading through the GitHub Runner code and depending on internal APIs that break without notice.

1. As you pointed out, this is documented multiple times in the README and TROUBLESHOOTING guide. We do our best to call this out.

1. In the case of Workload Identity Federation, the risk leak is fairly minimal. While the JSON file does contain a _GitHub_ OIDC token, the token is pinned to a workflow, and GitHub documents that the token expires when the workflow finishes. Therefore the surface area for an attack is quite small.

    For Service Account Key JSON, you could theoretically not use  this entire action and just write the JSON file to disk and set `$GOOGLE_APPLICATION_CREDENTIALS`.

1. Storing credentials in environment variables doesn't work. The Google Cloud SDKs require that credentials exist on a file on disk.

I understand the frustration and concern, but for all these reasons, I'm going to close this out.

**Comment by matthieuprat** [comment]
Hey @sethvargo, thank you for the quick and thorough response. It all makes sense.

Out of curiosity, if the `RUNNER_TEMP` directory became accessible to Docker-based actions, is using that directory something you'd consider again?

**Comment by sethvargo** [comment]
Hi @matthieuprat - if there were any well-established and well-documented directory that GitHub recommended, we would definitely consider it. But we'd need a stronger guarantee beyond "it exists in the source code".
